File: Unity/Factory.cs
using Logic;
using Infrastructure.Logic;
using Shared.Interfaces;
using Unity.Injection;
using Infrastructure.Data;
using Data.Repositories;

namespace Unity
{
    public class Factory : IFactory
    {
        private IUnityContainer _unit;
        public Factory()
        {
            _unit = new UnityContainer();
            Register();
        }

        private void Register()
        {
            _unit.RegisterType<IBookDM, BookDM>(new InjectionConstructor(this));
            _unit.RegisterType<IAuthorDM, AuthorDM>(new InjectionConstructor(this));

            _unit.RegisterType<IBookRepo, BookRepo>();
            _unit.RegisterType<IAuthorRepo, AuthorRepo>();
        }

        public T GetService<T>()
        {
            return (T)_unit.Resolve(typeof(T));
        }
    }
}

File: UnityVB/Factory.vb
Imports Data.Repositories
Imports Infrastructure.Data
Imports Infrastructure.Logic
Imports Logic
Imports [Shared].Interfaces
Imports Unity
Imports Unity.Injection

Public Class Factory
    Implements IFactory

    Private _unit As IUnityContainer

    Public Sub New()
        _unit = New UnityContainer()
        Register()
    End Sub

    Private Sub Register()
        _unit.RegisterType(Of IBookDM, BookDM)(New InjectionConstructor(Me))
        _unit.RegisterType(Of IAuthorDM, AuthorDM)(New InjectionConstructor(Me))

        _unit.RegisterType(Of IBookRepo, BookRepo)()
        _unit.RegisterType(Of IAuthorRepo, AuthorRepo)()
    End Sub

    Public Function GetService(Of T)() As T Implements IFactory.GetService
        Return _unit.Resolve(GetType(T))
    End Function


End Class